### ISSUE TITLE
[React-Native] Expo Web doesn't support `styled.Component` notation

### ISSUE THREAD

**Issue by danielkov** [styled-components/styled-components#2624]
## System:
 - OS: macOS 10.15
 - CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
 - Memory: 1.31 GB / 16.00 GB
 - Shell: 5.7.1 - /bin/zsh
## Binaries:
 - Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
 - Yarn: 1.15.2 - /usr/local/bin/yarn
 - npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
 - Watchman: 4.9.0 - /usr/local/bin/watchman
## npmPackages:
 - styled-components: ^4.3.1 => 4.3.1 

## Reproduction

1. Create a new project with Expo, using TypeScript.
2. Add `styled-components` as dependency.
3. The target `"web"` should be specified in `app.json`, like so:
```json
"platforms": [
  "ios",
  "android",
  "web"
],
```

## Steps to reproduce

1. Add the following line to any used file: `import styled from 'styled-components/native`.
2. Use `const Test = styled.View` somewhere.

## Expected Behavior

Module builds, everything compiles and I get my styled component.

## Actual Behavior

Module build warns (does not error for some reason): 
```
Attempted import error: 'react-native' does not contain a default export (imported as 'reactNative').
```
in
```
styled-components/native/dist/styled-components.native.esm.js
```

It works fine if I import the component separately and use `styled()` as a function, like so:

```jsx
import styled from 'styled-components';
import { View } from 'react-native';

const Test = styled(View) //...
```

**Comment by probablyup** [comment]
Can you try using the v5 beta? We reworked this to support RNW that may fix this as well?

**Comment by ruslan-mystore** [comment]
same problem here

**Comment by danielkov** [comment]
> Can you try using the v5 beta? We reworked this to support RNW that may fix this as well?

I will try and update you about this. I'll need to set the project up without TypeScript though, because last time I check the `@typings` repository was not updated for v5 beta.

**Comment by geelen** [comment]
Oooh so I was able to confirm that this is fixed in `v5` but _only_ if you `import styled from 'styled-components/native'` (not `import styled from 'styled-components'`)

<img width="426" alt="image" src="https://user-images.githubusercontent.com/23264/60150977-3a88de80-981e-11e9-8d3d-7dde05f6d002.png">
<img width="731" alt="image" src="https://user-images.githubusercontent.com/23264/60150994-4aa0be00-981e-11e9-8f50-9615d30a1254.png">

Same behaviour with `styled(Text)` with the same import.

With `import styled from 'styled-components'`, on Expo web you get the DOM version, which is still busted in the same way:

<img width="364" alt="image" src="https://user-images.githubusercontent.com/23264/60151027-82a80100-981e-11e9-934f-ff9ab1b7931d.png">
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/23264/60151031-88054b80-981e-11e9-828e-785aaf86d2ac.png">

Switching to `styled(Text)` "works", but treats the `Text` component like a web component, passing className to it, which ends up with wonky styles:

<img width="786" alt="image" src="https://user-images.githubusercontent.com/23264/60151078-bbe07100-981e-11e9-901a-cbeb471c63f3.png">

(Lol, just figured out that that was because I had `line-height: 24;` not `line-height: 24px;`. But that shows you about how the DOM and RN versions of SC differ, hey?)

You also get this warning:

<img width="454" alt="image" src="https://user-images.githubusercontent.com/23264/60151101-d9153f80-981e-11e9-8300-5388f3a48032.png">

Which is discussed in this RNW issue: https://github.com/necolas/react-native-web/issues/1146

As for `styled-components/native` vs `styled-components`, I couldn't remember when that changed. Turns out it was part of the [v3 release](https://github.com/styled-components/styled-components/releases/tag/v3.0.1) 18 months ago. I'm guessing anyone who's done the RNW conversion until this point has been accidentally injecting `className` props and things have been mostly ok, until now.

TL;DR: v5 works as long as you import `styled-components/native`!

**Comment by geelen** [comment]
Found another way, since _something_ we were doing in our app was conflicting with the hooks implementation of v5.

You can stub out the default export in RNW like this:

```js
// config-overrides.js from react-app-rewired

module.exports = function(config, env) {
  config.resolve = {
    ...config.resolve,
    alias: {
      ...config.resolve.alias,
      'react-native': path.resolve(__dirname, 'mocks/react-native')
    }
  };

  return config;
};
```

```js
// mocks/react-native.js
import * as RN from 'react-native-web';
export default RN;

export {
  createElement,
  findNodeHandle,
  render,
  unmountComponentAtNode, // modules
  processColor,
  NativeModules,
  TextPropTypes,
  ViewPropTypes, // APIs
  AccessibilityInfo,
  Alert,
  Animated,
  AppRegistry,
  AppState,
  AsyncStorage,
  BackHandler,
  Clipboard,
  DeviceInfo,
  Dimensions,
  Easing,
  I18nManager,
  InteractionManager,
  Keyboard,
  LayoutAnimation,
  Linking,
  NativeEventEmitter,
  NetInfo,
  PanResponder,
  PixelRatio,
  Platform,
  Share,
  StyleSheet,
  UIManager,
  Vibration, // components
  ActivityIndicator,
  Button,
  CheckBox,
  FlatList,
  Image,
  ImageBackground,
  KeyboardAvoidingView,
  ListView,
  Modal,
  Picker,
  ProgressBar,
  RefreshControl,
  SafeAreaView,
  ScrollView,
  SectionList,
  Slider,
  StatusBar,
  SwipeableFlatList,
  SwipeableListView,
  Switch,
  Text,
  TextInput,
  Touchable,
  TouchableHighlight,
  TouchableNativeFeedback,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
  VirtualizedList,
  YellowBox, // propTypes
  ColorPropType,
  EdgeInsetsPropType,
  PointPropType, // compat (components)
  DatePickerIOS,
  DrawerLayoutAndroid,
  ImageEditor,
  ImageStore,
  InputAccessoryView,
  MaskedViewIOS,
  NavigatorIOS,
  PickerIOS,
  ProgressBarAndroid,
  ProgressViewIOS,
  SegmentedControlIOS,
  SnapshotViewIOS,
  TabBarIOS,
  ToastAndroid,
  ToolbarAndroid,
  ViewPagerAndroid,
  WebView, // compat (apis)
  ActionSheetIOS,
  AlertIOS,
  CameraRoll,
  DatePickerAndroid,
  ImagePickerIOS,
  PermissionsAndroid,
  PushNotificationIOS,
  Settings,
  StatusBarIOS,
  Systrace,
  TimePickerAndroid,
  TVEventHandler,
  VibrationIOS
} from 'react-native-web';
```

Ain't pretty, but it works. Gets me styled-components v4.3.2 working with react-native-web v0.11.4

**Comment by Aleksion** [comment]
I'm facing some of these issues with NextJS. It only happens with styled-components. Has anyone else hit this?

The issues SEEMS to be something related to improper aliasing of "react-native" from within styled-components. I'm unsure how to solve it though.

I know how to make it work with Razzle.js, but I can't for the life of me figure out how to get to the same working state with NextJS. My suspicion is that it has to do with webpack's config.externals - BUT it could also be something in babel.config.js.

If someone has tackled this issue, you'll be my favourite person of the year.
I've setup to repo's reproducing the issue

--- Next.js
pages/index.js - WORKS
pages/problem.js - FAILS (has styled-components/native)

--- Razzle
pages/Home.js - WORKS
pages/Home.js - WORKS (has styled-components/native)

https://github.com/Aleksion/rnw-styled-next
https://github.com/Aleksion/rnw-styled-razzle

**Comment by elie222** [comment]
Recommend trying this fix:
https://github.com/styled-components/styled-components/pull/2797

Worked for me.
